x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java

```diff
@@ -121,6 +150,71 @@ protected void onStart(long now, ActionListener<Void> listener) {
 
     @Override
     protected SearchRequest buildSearchRequest() {
-        return pivot.buildSearchRequest(getPosition());
+        return pivot.buildSearchRequest(getPosition(), pageSize);
+    }
+
+    /**
+     * Handle the circuit breaking case: A search consumed to much memory and got aborted.
+     *
+     * Going out of memory we smoothly reduce the page size which reduces memory consumption.
+     *
+     * Implementation details: We take the values from the circuit breaker as a hint, but
+     * note that it breaks early, that's why we also reduce using
+     *
+     * @param e Exception thrown, only {@link CircuitBreakingException} are handled
+     * @return true if exception was handled, false if not
+     */
+    protected boolean handleCircuitBreakingException(Exception e) {
+        CircuitBreakingException circuitBreakingException = getCircuitBreakingException(e);
+
+        if (circuitBreakingException == null) {
+            return false;
+        }
+
+        double reducingFactor = Math.min((double) circuitBreakingException.getByteLimit() / circuitBreakingException.getBytesWanted(),
+                1 - (Math.log10(pageSize) * 0.1));
+
+        int newPageSize = (int) Math.round(reducingFactor * pageSize);
```

Comment:
```suggestion
        int newPageSize = Math.min((int) Math.round(reducingFactor * pageSize), MINIMUM_PAGE_SIZE);
```
I think doing this is best, in conjunction with having a limit on the number of times we attempt a search again. I don't think we should limit our number attempts based on the gradient descent of our page size. It seems to me that it should be a static number. As you pointed out yourself, sometimes just trying again with the same page size works.

Reply:
that change would immediately set the page size to `MINIMUM_PAGE_SIZE` no matter what.

The number of retries is limited: if the page size gets below `MINIMUM_PAGE_SIZE` the handler fails and therefore ends in the existing error handler, which eventually sets the job to `FAILED`. I have tested that it isn't to smooth to avoid a endless loop.

Reply:
as for retry: Note that on a circuit break the job is put back to `STARTED`, because that's what `onFailure` does.

I considered intercepting the response instead, to not get into the failure handler, but decided against it. Once we reworked the scheduler we should revisit, it would be good to wait a bit till the next search attempt.

Reply:
woops, I meant `max` :D

Reply:
the `if` on the next line handles this case, but I just realize `MINIMUM_PAGE_SIZE` is off by 1